### PR TITLE
Don't persist keys for embedded entities

### LINKED ISSUE #13578: Cosmos: Preserve unmapped properties for owned collections
Currently when updating an entity that has a nested owned collection the unmapped values will be lost.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
The store JObject is now stored in a shadow property for embedded entities as well to allow to keep unmapped properties.

Fixes #13578
Part of #12086

### CODE REVIEW

**Comment by AndriySvyryd** [comment]
This is ready for review now

File: src/EFCore.Cosmos/Extensions/CosmosPropertyExtensions.cs [hunk 1]
```diff
@@ -20,7 +21,27 @@ public static class CosmosPropertyExtensions
...
+
+            if (ownership != null
+                && !entityType.IsDocumentRoot())
+            {
+                var pk = property.FindContainingPrimaryKey();
+                if (pk != null
+                    && pk.Properties.Count == ownership.Properties.Count + (ownership.IsUnique ? 0 : 1)
+                    && ownership.Properties.All(fkProperty => pk.Properties.Contains(fkProperty)))
+                {
+                    return "";
```
Comment (ajcvickers) [inline src/EFCore.Cosmos/Extensions/CosmosPropertyExtensions.cs]: Empty string?

Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Extensions/CosmosPropertyExtensions.cs]: Yes, that's the way to mark a property as not persisted until https://github.com/aspnet/EntityFrameworkCore/issues/14121 is implemented.

**Comment by AndriySvyryd** [comment]
@smitpatel Please review this when you are able

File: src/EFCore.Cosmos/Query/Internal/EntityProjectionExpression.cs [hunk 2]
```diff
@@ -156,7 +170,7 @@ public virtual Expression BindNavigation(INavigation navigation)
         ///     any release. You should only use it directly in your code with extreme caution and knowing that
         ///     doing so can result in application failures when updating to a new Entity Framework Core release.
         /// </summary>
-        public virtual Expression BindMember(string name, Type entityClrType, out IPropertyBase propertyBase)
+        public virtual Expression BindMember(string name, Type entityClrType, bool clientEval, out IPropertyBase propertyBase)
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Query/Internal/EntityProjectionExpression.cs]: Could DRY if using MemberIdentity

File: src/EFCore.Cosmos/Update/Internal/DocumentSource.cs [hunk 3]
```diff
@@ -143,62 +153,71 @@ public virtual JObject UpdateDocument(JObject document, IUpdateEntry entry)
...
                     }
                 }
                 else if (fk.IsUnique)
                 {
-                    var nestedEntry = ((InternalEntityEntry)entry).StateManager.TryGetEntry(nestedValue, fk.DeclaringEntityType);
-                    if (nestedEntry == null)
+                    var embeddedEntry = ((InternalEntityEntry)entry).StateManager.TryGetEntry(embeddedValue, fk.DeclaringEntityType);
+                    if (embeddedEntry == null)
                     {
                         return document;
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Update/Internal/DocumentSource.cs]: return or continue? It is inside a foreach loop

File: src/EFCore.Cosmos/Update/Internal/DocumentSource.cs [hunk 4]
```diff
@@ -143,62 +153,71 @@ public virtual JObject UpdateDocument(JObject document, IUpdateEntry entry)
...
+                        var embeddedEntry = ((InternalEntityEntry)entry).StateManager.TryGetEntry(dependent, fk.DeclaringEntityType);
+                        if (embeddedEntry == null)
                         {
                             continue;
                         }
 
-                        array.Add(_database.GetDocumentSource(dependentEntry.EntityType).CreateDocument(dependentEntry));
+                        var embeddedDocument = embeddedDocumentSource.GetCurrentDocument(embeddedEntry);
+                        embeddedDocument = embeddedDocument != null
+                            ? embeddedDocumentSource.UpdateDocument(embeddedDocument, embeddedEntry) ?? embeddedDocument
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Update/Internal/DocumentSource.cs]: Why this uses coalesce but reference navigation is not using?

File: src/EFCore.Cosmos/Update/Internal/DocumentSource.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Update/Internal/DocumentSource.cs]: Nice catch

File: src/EFCore.Cosmos/Update/Internal/DocumentSource.cs (hunk 4 above)
Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Update/Internal/DocumentSource.cs]: We need to add all entities to the collection in order even if they weren't updated

**Pull request merged**